File: Leetcode/Algorithm/python/0000/00750-Number Of Corner Rectangles.py
class Solution(object):
    def countCornerRectangles(self, grid):
        n = len(grid)
        if n < 2:
            return 0
        m = len(grid[0])
        if m < 2:
            return 0
        
        d = [ set() for i in xrange(n) ]
        
        for i in xrange(n):
            d[i] = set([j for (j, v) in enumerate(grid[i]) if v == 1])
        
        res = 0
        for i in xrange(n):
            for j in xrange(i + 1, n):
                u = len(d[i] & d[j])
                res += u * (u - 1) / 2
                        
        return res

File: Leetcode/Algorithm/cpp/0000/00750-Number Of Corner Rectangles.cc
class Solution {
public:
    int countCornerRectangles(vector<vector<int>>& grid) {
        vector<pair<int, int> > points;
        int n = grid.size();
        int m = grid[0].size();
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 0) {
                    continue;
                }
                points.push_back({i, j});
            }
        }
        
        int res = 0;
        int k = points.size();
        for (int i = 0; i < k; i++) {
            for (int j = i + 1; j < k; j++) {
                int y1 = points[i].first;
                int x1 = points[i].second;
                int y2 = points[j].first;
                int x2 = points[j].second;
                
                if (y1 == y2 || x1 == x2) {
                    continue;
                }
                
                if (grid[y1][x2] && grid[y2][x1]) {
                    res++;
                }
            }
        }
        return res / 2;
    }
};
